src/PowerShellEditorServices/Services/PowerShell/Console/PSReadLineProxy.cs

```diff
@@ -0,0 +1,198 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Management.Automation;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Microsoft.Extensions.Logging;
+using Microsoft.PowerShell.EditorServices.Services.PowerShell.Utility;
+using SMA = System.Management.Automation;
+
+namespace Microsoft.PowerShell.EditorServices.Services.PowerShell.Console
+{
+    using System.Management.Automation.Runspaces;
+
+    internal class PSReadLineProxy
+    {
+        private const string FieldMemberType = "field";
+
+        private const string MethodMemberType = "method";
+
+        private const string AddToHistoryMethodName = "AddToHistory";
+
+        private const string SetKeyHandlerMethodName = "SetKeyHandler";
+
+        private const string ReadLineMethodName = "ReadLine";
+
+        private const string ReadKeyOverrideFieldName = "_readKeyOverride";
+
+        private const string HandleIdleOverrideName = "_handleIdleOverride";
+
+        private const string VirtualTerminalTypeName = "Microsoft.PowerShell.Internal.VirtualTerminal";
+
+        private const string ForcePSEventHandlingMethodName = "ForcePSEventHandling";
+
+        private static readonly Type[] s_setKeyHandlerTypes =
+        {
+            typeof(string[]),
+            typeof(Action<ConsoleKeyInfo?, object>),
+            typeof(string),
+            typeof(string)
+        };
+
+        private static readonly Type[] s_addToHistoryTypes = { typeof(string) };
+
+        private static readonly string _psReadLineModulePath = Path.GetFullPath(
+            Path.Combine(
+                Path.GetDirectoryName(typeof(PSReadLineProxy).Assembly.Location),
+                "..",
+                "..",
+                "..",
+                "PSReadLine"));
+
+        private static readonly string ReadLineInitScript = $@"
+            [System.Diagnostics.DebuggerHidden()]
+            [System.Diagnostics.DebuggerStepThrough()]
+            param()
+            end {{
+                $module = Get-Module -ListAvailable PSReadLine |
+                    Where-Object {{ $_.Version -ge '2.0.2' }} |
+                    Sort-Object -Descending Version |
+                    Select-Object -First 1
+                if (-not $module) {{
+                    Import-Module '{_psReadLineModulePath.Replace("'", "''")}'
+                    return [Microsoft.PowerShell.PSConsoleReadLine]
+                }}
+
+                Import-Module -ModuleInfo $module
+                return [Microsoft.PowerShell.PSConsoleReadLine]
```

Comment:
😱

Reply:
We should both stop supporting loading any arbitrary PSReadLine in and only support what we ship, and ideally we should also pull it in as a .NET dependency instead of a PowerShell module (though that comes with its own complications).

Reply:
/cc @daxian-dbw